Test plugins with pyln-testing and travis

cdecker opened the pull request:
Running tests in Travis is a nice safety net, so why not :-)

m-schmoock commented:
Also, we could need a `README.md` on how to run plugin tests locally manually. If you do this, please also change the outdated info ion section `### Pylightning` `PYTHONPATH` missing the pyln modules.

cdecker commented on .travis/build.sh:
```diff
@@ -0,0 +1,48 @@
+    rm -rf bitcoin-0.17.1-x86_64-linux-gnu.tar.gz bitcoin-0.17.1
+fi
+
+pyenv global 3.7.1
```
Well, this is supposed to run inside a travis docker image, which does have quite a bit of extra stuff. This command just pins the python version to a specific release, instead of whatever the default was.

cdecker commented on .travis/build.sh:
```diff
@@ -0,0 +1,48 @@
+export LIGHTNING_VERSION=${LIGHTNING_VERSION:-master}
+export PYTHONPATH=/tmp/lightning/contrib/pyln-client:/tmp/lightning/contrib/pyln-testing:/tmp/lightning/contrib/pylightning:$$PYTHONPATH
+
+mkdir -p dependencies/bin || true
```
Oh, I wasn't aware that `-p` had this guarantee. Will remove :+1:

cdecker commented:
Fixed the things pointed out by @m-schmoock and rewrote part of the readme. I thought it might be useful to test against the full configuration matrix, including `COMPAT`, `DEVELOPER` and `EXPERIMENTAL_FEATURES` flags so plugin authors can see issues early on.

Also travis is configured to run once a day against c-lightning `master` so we should see if a change in  c-ligthning breaks things.

m-schmoock commented on README.md:
```diff
@@ -45,19 +46,45 @@ all plugin directories into your `~/.lightning/plugins` directory. The daemon
+importantly that they will continue to work with newly release versions of
+c-lightning.
+
+Writing a test is as simple as this:
```
Maybe also point out that the framework will look for files starting with "test_" and functions with "test_" in those files.

```
To write a test, simply add a python file starting with 'test_' in your plugin's directory and add the test code in a function also starting with 'test_' like this:
```

m-schmoock replied:
Also, an instruction on howto run the tests locally before committing, would be great (if possible). I fiddled around with `pytest` and PYTHONPATH myself, but it does not launch the regtest daemons and thus runs into connection errors.

m-schmoock commented:
@cdecker in the travis logs there is a lot of:
`pyln-testing 0.0.1 has requirement pytest==5.0.1, but you'll have pytest 5.3.1 which is incompatible.`, can we ignore this safely?

cdecker commented:
> @cdecker in the travis logs there is a lot of:
> `pyln-testing 0.0.1 has requirement pytest==5.0.1, but you'll have pytest 5.3.1 which is incompatible.`, can we ignore this safely?

Yes we can, they are just requiring older versions. Technically this would be an issue, but it's not really. I'll make sure we at least sync up the dependencies between pyln-testing and this repo.

cdecker commented:
> > @cdecker in the travis logs there is a lot of:
> > `pyln-testing 0.0.1 has requirement pytest==5.0.1, but you'll have pytest 5.3.1 which is incompatible.`, can we ignore this safely?
> 
> Yes we can, they are just requiring older versions. Technically this would be an issue, but it's not really. I'll make sure we at least sync up the dependencies between pyln-testing and this repo.

I added `pyln-testing` as an early dependency so it now dictates which `pytest` version we install :-)

I also added an upgrade to `pip` so it doesn't keep complaining about old versions.

cdecker commented:
If we wanted to push this to the limit we could create a separate `virtualenv` for each plugin and install the specific versions the plugin is asking for, but that's for another time I think :-)
